pkg/v2/tkr/resolver/internal/resolver_impl.go

```diff
@@ -0,0 +1,366 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package internal contains the implementation of the TKR Resolver.
+package internal
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"sigs.k8s.io/cluster-api/util/conditions"
+
+	runv1 "github.com/vmware-tanzu/tanzu-framework/apis/run/v1alpha3"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v2/tkr/resolver/data"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v2/tkr/util/version"
+)
+
+type Resolver struct {
+	cache cache
+}
+
+func NewResolver() *Resolver {
+	return &Resolver{cache: cache{
+		tkrs:                 data.TKRs{},
+		osImages:             data.OSImages{},
+		tkrsShippingOSImage:  map[string]data.TKRs{},
+		osImagesShippedByTKR: map[string]data.OSImages{},
+	}}
+}
+
+// Add pre-validated TKRs and OSImages to the Resolver cache.
+// Pre-req: invalid TKRs and OSImages MUST have status condition Valid=False.
+// Invalid objects would have Kubernetes versions inconsistent between a TKR and its OSImages, TKR and Kubernetes
+// versions that would not parse successfully.
+func (r *Resolver) Add(objects ...interface{}) {
+	r.cache.add(objects)
+}
+
+func (r *Resolver) Resolve(query data.Query) data.Result {
+	query = normalize(query)
+
+	result := r.cache.filter(query)
+	return r.cache.sort(result)
+}
+
+type osImageDetails struct {
+	tkrs          data.TKRs
+	osImagesByTKR map[string]data.OSImages
+}
+
+type cache struct {
+	mutex sync.RWMutex
+
+	tkrs     data.TKRs
+	osImages data.OSImages
+
+	// indices
+	tkrsShippingOSImage  map[string]data.TKRs
+	osImagesShippedByTKR map[string]data.OSImages
+}
+
+type details struct {
+	controlPlane       osImageDetails
+	machineDeployments map[string]osImageDetails
+}
+
+func (cache *cache) add(objects []interface{}) {
+	cache.mutex.Lock()
+	defer cache.mutex.Unlock()
+
+	for _, object := range objects {
+		cache.addObject(object)
+	}
+}
+
+func (cache *cache) addObject(object interface{}) {
+	switch object := object.(type) {
+	case *runv1.TanzuKubernetesRelease:
+		if !object.DeletionTimestamp.IsZero() {
```

Comment:
I kinda understand why this is being done but deleting from cache on add is counterintuitive and is a symptom that a background cache refresh is needed.

Reply:
Looks like Cache.Delete() method wouldn't hurt, for clarity. TODO

Reply:
Added `Remove()` method to Cache.